src/dbnode/storage/index/types.go

```diff
@@ -125,16 +135,108 @@ type ResultsAllocator func() Results
 
 // ResultsPool allows users to pool `Results` types.
 type ResultsPool interface {
-	// Init initialized the results pool.
+	// Init initializes the results pool.
 	Init(alloc ResultsAllocator)
 
 	// Get retrieves a Results object for use.
 	Get() Results
 
-	// Put returns the provide value to the pool.
+	// Put returns the provided Results to the pool.
 	Put(value Results)
 }
 
+// AggregateResults is a collection of results for an aggregation query.
+type AggregateResults interface {
+	// Namespace returns the namespace associated with the result.
+	Namespace() ident.ID
+
+	// Map returns a map from tag name -> possible tag values,
+	// comprising search results.
+	Map() *AggregateResultsMap
+
+	// Reset resets the AggregateResults object to initial state.
+	Reset(nsID ident.ID)
+
+	// Finalize releases any resources held by the AggregateResults object,
+	// including returning it to a backing pool.
+	Finalize()
+
+	// NoFinalize marks the AggregateResults such that a subsequent call to
+	// Finalize() will be a no-op and will not return the object to the pool or
+	// release any of its resources.
+	NoFinalize()
+
+	// Size returns the number of results discovered.
+	Size() int
+
+	// AggregateDocument converts the provided document to a set of tags
+	// fulfilling the aggregate query and adds it to the results. This method makes
+	// a copy of the bytes backing the tags (TODO: does it?), so the original may
+	// be modified after this function returns without affecting the results map.
+	AggregateDocument(document doc.Document, opts QueryOptions) error
+
+	// AddIDAndValues adds  the given values to the given ID set.
+	//
+	// NB: this does not need to take in query options for filtering, since all
+	// incoming values are presumed to have already been filtered.
+	AddIDAndValues(id ident.ID, values AggregateValues) error
+}
+
+// AggregateResultsAllocator allocates AggregateResults types.
+type AggregateResultsAllocator func() AggregateResults
+
+// AggregateResultsPool allows users to pool `AggregateResults` types.
+type AggregateResultsPool interface {
+	// Init initializes the AggregateResults pool.
+	Init(alloc AggregateResultsAllocator)
+
+	// Get retrieves a AggregateResults object for use.
+	Get() AggregateResults
+
+	// Put returns the provided AggregateResults to the pool.
+	Put(value AggregateResults)
+}
+
+// AggregateValues is a collection of values for an aggregation query.
+type AggregateValues interface {
```

Comment:
is this just a wrapper on a code-gen'd type, if so - can we use the struct where needed instead?

Reply:
Good call, I'll give it a shot, hopefully the code gen will still be happy with it